### PR TITLE
Container component to dynamically replace an async loaded Drawable

### PR DESCRIPTION
**Pull request by swoolcock**
Will be used in osu! for things like user avatar and cover background.
Related to https://github.com/ppy/osu/issues/2445

~~Note there is currently an edge case where spamming updates can occasionally leave `DelayedLoadWrapper`s that do not expire.  I welcome comments and/or solutions.~~
Fixed.

### CODE REVIEW

File: osu.Framework.Tests/Visual/TestCaseUpdateableContainer.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            {
+                Position = new Vector2(50, 50);
+                Text = item == null ? "No Item" : $"Item {item.ItemId}";
+            }
+
+            [BackgroundDependencyLoader]
+            private void load()
+            {
+                // delay
+                Thread.Sleep(500);
```
Comment (peppy) [inline osu.Framework.Tests/Visual/TestCaseUpdateableContainer.cs]: This can be incompatible with headless execution, as rate of game clock's execution is increased. I can't find another example of how this is done currently, but something to consider

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        /// </summary>
+        /// <param name="item">The source item that the Drawable should represent.</param>
+        protected virtual Drawable CreateDrawable(T item) => null;
+
+        /// <summary>
+        /// Override to perform a custom comparison of two source items. By default an object reference comparison
+        /// is used, but it may be desirable to compare based on properties of the items.
+        /// </summary>
+        /// <returns><c>true</c>, if the items are logically equivalent, <c>false</c> otherwise.</returns>
+        protected virtual bool CompareItems(T lhs, T rhs) => lhs == rhs;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: is there a valid use case for this that you had when implementing it?

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        /// </summary>
+        protected virtual Drawable CreatePlaceholder() => null;
+
+        /// <summary>
+        /// Override to instantiate a custom Drawable based on the passed source item.
+        /// May be null to indicate that the source item has no visual representation,
+        /// in which case the placeholder will be used if it exists.
+        /// </summary>
+        /// <param name="item">The source item that the Drawable should represent.</param>
+        protected virtual Drawable CreateDrawable(T item) => null;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: should probably be `abstract`

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            }
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            updateDrawable();
+        }
+
+        private void updateDrawable()
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: this method is too long and not customisable.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+            if (newDrawable == DisplayedDrawable)
+                return;
+
+            NextDrawable = newDrawable;
+
+            if (newDrawable == null && DisplayedDrawable != null)
+            {
+                DisplayedDrawable?.FadeOut(300).Expire();
+                PlaceholderDrawable?.FadeInFromZero(300, Easing.OutQuint);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: new method `virtual ShowDrawable(Drawable d)` which is used for both fading in the placeholder and the non-placeholder.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            var newDrawable = CreateDrawable(source);
+
+            if (newDrawable == DisplayedDrawable)
+                return;
+
+            NextDrawable = newDrawable;
+
+            if (newDrawable == null && DisplayedDrawable != null)
+            {
+                DisplayedDrawable?.FadeOut(300).Expire();
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: new method `HideDrawable(Drawable d)` used for all hiding. keep the expire logic local since the placeholder can't be expired.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                PlaceholderDrawable?.FadeInFromZero(300, Easing.OutQuint);
+                DisplayedDrawable = null;
+                return;
+            }
+
+            if (FadeOutImmediately)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: move above early return block and change conditional to `if (newDrawable == null || FadeOutImmediately)` to reduce code duplication.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs (hunk 6 above)
Comment (peppy) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: also 300 should be a `protected virtual double FadeDuration => 300` to allow customisation

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    if (!FadeOutImmediately)
+                        DisplayedDrawable?.FadeOut(300).Expire();
+
+                    d.FadeInFromZero(300, Easing.OutQuint);
+
+                    DisplayedDrawable = d;
+                    NextDrawable = null;
+                };
+
+                Add(new DelayedLoadWrapper(newDrawable));
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: no customisation is given over the time before load variable, which seems important.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    /// </summary>
+    public abstract class UpdateableContainer<T> : Container where T : class
+    {
+        /// <summary>
+        /// The placeholder Drawable created when the container was instantiated.
+        /// </summary>
+        public readonly Drawable PlaceholderDrawable;
+
+        /// <summary>
+        /// The currently displayed Drawable. Null if no Drawable or the placeholder is displayed.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: This reads awkwardly. I'd probably rephrase to `Null if no drawable is displayed (note that the placeholder may still be displayed in this state).`

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs (hunk 2 above)
Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: `User` needs to be compared by `Id` rather than by reference, in case its details have not been fully populated.

Comment (peppy) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: i see. probably want to use `IComparer<T>` here (see `SortedList` for an example)

Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Does this implementation sound ok?
`protected virtual IComparer<T> Comparer => Comparer<T>.Default;`

Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Potentially creating a new `Comparer` every time though, depending on how subclasses implement it.

**Comment by peppy** [comment]
image

isn't this wrong? seems to be missing asserts. also you probably want to add a test for other features like the delayed load.

**Comment by swoolcock** [comment]
Hmm that test used to work, might be a regression when I reimplemented `updateDrawable`.  I'll look into it.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    /// <summary>
+    /// Manages dynamically displaying a custom Drawable based on a "source" object.
+    /// Useful for replacing Drawables on the fly.
+    /// </summary>
+    public abstract class UpdateableContainer<T> : Container where T : class
+    {
+        /// <summary>
+        /// The placeholder Drawable created when the container was instantiated.
+        /// </summary>
+        public readonly Drawable PlaceholderDrawable;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: I don't think these need to be `public`. Best to make them protected (or private where possible) and add a test-specific class to access them.

(`PlaceholderDrawable`, `DisplayedDrawable `, `NextDrawable`, `LoadTriggered`, `Comparer`)

Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Hmm I thought they might be useful to developers.  Ok, will change.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+using System.Collections.Generic;
+using osu.Framework.Lists;
+
+namespace osu.Framework.Graphics.Containers
+{
+    /// <summary>
+    /// Manages dynamically displaying a custom Drawable based on a "source" object.
+    /// Useful for replacing Drawables on the fly.
+    /// </summary>
+    public abstract class UpdateableContainer<T> : Container where T : class
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: This should be a `CompositeDrawable`, not a `Container`.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+        /// <summary>
+        /// The currently displayed Drawable. Null if no drawable is displayed (note that the placeholder may still be displayed in this state).
+        /// </summary>
+        public Drawable DisplayedDrawable { get; private set; }
+
+        /// <summary>
+        /// The Drawable that will be presented next. Null if there is no pending drawable.
+        /// </summary>
+        public Drawable NextDrawable { get; private set; }
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/UpdateableContainer.cs]: I was thinking that we should be able to blackbox test these values by observing the state drawables within the testcase, rather than exposing them ourselves, and the placeholder. Would really prefer they were private, as stated by @peppy .

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs (hunk 11 above)
Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Ah right, I was basing it on `UpdateableAvatar`, which is a `Container`.

Comment (peppy) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: How about `ModelBackedDrawable`?

Comment (smoogipoo) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Let's go with `ModelBackedDrawable` instead. Don't want to mentioned "updateable" since that ties into the whole update hierarchy of drawables.

Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: Rename `Source` to `Model` too?

Comment (smoogipoo) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: I was okay with renaming `Source` to `Model` in the first place.

File: osu.Framework/Graphics/Containers/UpdateableContainer.cs (hunk 10 above)
Comment (swoolcock) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: What about making some of them `internal`?  That way they're not exposed outside the framework, but the visual tests can still access them.

Comment (smoogipoo) [reply osu.Framework/Graphics/Containers/UpdateableContainer.cs]: There should be very minimal coupling between the osu.Game and osu.Game.Tests assemblies. The test case should not have to be changed if the underlying logic of `UpdateableContainer` is changed.

Disagree with making internal.

**Comment by swoolcock** [comment]
I've added a `CreateDelayedLoadWrapper` method that can be overridden by subclasses to provide a custom `DelayedLoadWrapper`, but that means I can no longer expose `LoadTriggered` from a private subclass.

I'm not sure whether `LoadTriggered` would actually be useful for `ModelBackedDrawable`, but regardless I need a way to verify in the visual test that `LoadDelay` is doing what it's supposed to.

Any suggestions?

**Comment by swoolcock** [comment]
Note that changing this to a `CompositeDrawable` means that `UpdateableAvatar` etc. in osu! will now need to be wrapped in a `Container` to apply the masking and edge effects.

**Comment by peppy** [comment]
image

I don't think this should happen.

**Comment by peppy** [comment]
A common scenario of use would be to always fade the new element above the previous one (and expire the previous one after the fade completes) rather than calling `FadeOut` on the old one. This avoids a period where both opacities are below 1.0 which produces a bad visual effect.

I'd go as far as saying thing should probably be default behaviour.

That said, it doesn't work correctly at the moment when a placeholder is being used, as the depth of the placeholder is always below every other added drawable. So we probably need depth management (incremented each time a drawable or placeholder is displayed).

**Comment by swoolcock** [comment]
`HideDrawable` accepting a null is intentional, since it doesn't check whether the placeholder exists when calling it.  I could do a check first instead?

**Comment by peppy** [comment]
It should definitely not be fired with a null drawable unless it is documented and has a reason to occur. I'd say just stop it from happening.

**Comment by swoolcock** [comment]
Once this one is merged I can push through another PR to resolve https://github.com/ppy/osu/issues/2445

**Comment by peppy** [comment]
I think this looks good to go. May evolve as we use it, but I'm happy with the code for now 👍
